lib/Dialect/Handshake/HandshakeTypes.cpp

```diff
@@ -99,9 +103,153 @@ checkChannelExtra(function_ref<InFlightDiagnostic()> emitError,
+/// This function is common to both ControlType and ChannelType.
+static LogicalResult
+parseExtraSignals(function_ref<InFlightDiagnostic()> emitError,
+                  AsmParser &odsParser,
+                  SmallVector<ExtraSignal::Storage> &extraSignalsStorage) {
+
+  auto parseSignal = [&]() -> ParseResult {
+    auto &signal = extraSignalsStorage.emplace_back();
+
+    if (odsParser.parseKeywordOrString(&signal.name) ||
+        odsParser.parseColon() || odsParser.parseType(signal.type))
+      return failure();
+
+    // Attempt to parse the optional upstream symbol
+    if (!odsParser.parseOptionalLParen()) {
+      std::string upstreamSymbol;
+      if (odsParser.parseKeywordOrString(&upstreamSymbol) ||
+          upstreamSymbol != UPSTREAM_SYMBOL || odsParser.parseRParen())
+        return failure();
+      // The default value of `signal.downstream` is true.
+      // I think this is a bit confusing.
+      // TODO: signal.upstream?
+      signal.downstream = false;
+    }
+    return success();
+  };
+
+  if (odsParser.parseCommaSeparatedList(parseSignal))
+    return failure();
+
+  return success();
+}
+
+//===----------------------------------------------------------------------===//
+// ControlType
+//===----------------------------------------------------------------------===//
+
+void ControlType::print(AsmPrinter &odsPrinter) const {
+  odsPrinter << "<";
+  if (!getExtraSignals().empty()) {
+    printExtraSignals(odsPrinter, getExtraSignals());
+  }
+  odsPrinter << ">";
+}
+
+LogicalResult ControlType::verify(function_ref<InFlightDiagnostic()> emitError,
+                                  ArrayRef<ExtraSignal> extraSignals) {
+  return checkChannelExtra(emitError, extraSignals);
+}
+
+/// Parse the control type after "<[" is parsed.
+static Type parseControlAfterLSquare(AsmParser &odsParser) {
+  function_ref<InFlightDiagnostic()> emitError = [&]() {
+    return odsParser.emitError(odsParser.getCurrentLocation());
+  };
+
+  SmallVector<ExtraSignal::Storage> extraSignalsStorage;
+  if (failed(parseExtraSignals(emitError, odsParser, extraSignalsStorage)))
+    return {};
+
+  // Parse ']'
+  if (odsParser.parseRSquare())
+    return {};
+  // Parse literal '>'
+  if (odsParser.parseGreater())
+    return {};
+
+  // We convert the storage type to the non-storage type at the end of parsing.
+  // Be careful that the non-storage type just holds a raw string pointer inside
+  // (via StringRef), so passing it around may lead to dangling pointers.
+  SmallVector<ExtraSignal> extraSignals;
+  // Convert the element type of the extra signal storage list to its
+  // non-storage version (these will be uniqued/allocated by ChannelType::get)
+  for (const ExtraSignal::Storage &signalStorage : extraSignalsStorage)
+    extraSignals.emplace_back(signalStorage);
+
+  if (failed(checkChannelExtra(emitError, extraSignals)))
+    return {};
+
+  return ControlType::get(odsParser.getContext(), extraSignals);
```

Comment:
add comment:

Build ControlType
Copies parsed data into MLIR context

Reply:
To me these two comments are useless and clutter the code. We are not supposed to re-document MLIR's inner workings here, they are abstracted away on purpose.

Reply:
I was discussing this with Shun the other day when we were talking about how to write comments. In general it seems to me your intended audience when you write comments is yourself- what I care about is when Aya is going to have a master's student who opens this file and is trying to understand what each line of code does.

Reply:
We can't possibly hope to comment what each LoC does, to me this is neither helpful nor desirable, even for new compiler devs. When the code is well-written, what each line does should be fairly obvious to anyone who can read C++ (we also can't hope to teach C++, this is a compiler not a tutorial). For example, if you have a function call, the function tells you what it does through its name, signature, and documentation. If the code is unclear, the solution is not to document every last line, it is to make the code clear.

I am all for having more larger comments explaining _why_ we do things the way we do them, which is actually helpful for any new pair of eyes, including more experienced developers that have never looked at a particular piece of code: _why_ a non-trivial design decision was made, _why_ do we use this data-structure when another simpler one would seem to work, etc. Commenting on _what_ the code does is helpful too when it gets complex (e.g., a for loop that is 50 lines long), but telling me that `ControlType::get` is essentially getting me a type is pointless. In fact, all the details about uniquing/allocating are leaving me wondering whether _I_ should care about memory management, when as the person writing this piece of code I should really really not. In a way this comment is puncturing a layer of abstraction that is there to make our lives easier.

Finally, fine-grained comments like this---in addition to cluttering the code---are a recipe for comment staleness i.e., comments going out of sync with the actual code, which is extra confusing for everyone. These are [standard recommendations](https://llvm.org/docs/CodingStandards.html#commenting) that many successful projects follow. The LLVM docs may use slightly different wording from me (what/how/etc.) but the message should be the same.

Reply:
>  In fact, all the details about uniquing/allocating are leaving me wondering whether I should care about memory management, when as the person writing this piece of code I should really really not. In a way this comment is puncturing a layer of abstraction that is there to make our lives easier.

This is basically the point on which we disagree- Shun was reading this code because there was a memory allocation issue. This is the section of the code that declares the StringRef, and points it to an std::string. However, that's not the std::string it stays pointing to, because MLIR moves the pointer. 

Additionally, the point of the comment being "Builds ControlType" is because the function is called get- it is not obvious from reading the code that this is a constructor.

Reply:
> This is the section of the code that declares the StringRef, and points it to an std::string. However, that's not the std::string it stays pointing to, because MLIR moves the pointer.

And in the end correct usage of the API results in correct behavior AFAIU, nothing surprising there, so no need to point it out in a comment. I don't know what was not working before, but if there is a general point someone wants to make about how (not) to manage memory around the creation of type/attributes then it should be made much clearer than what it is now and put in a more visible place (e.g., the function's doc), because as is I really don't know what people will get from this. If the problem was a simple mismanagement of memory like it can happen literally everywhere (which it sounds like it is), then there is really no point in leaving a note.

In any case, the people who will need to actually care about how MLIR manages memory should not understand it from this piece of code, they should go look at the official docs online or figure it out by following references from `ControlType::get`.

> it is not obvious from reading the code that this is a constructor.

Well it certainly is not just a constructor in the C++ sense. This behavior is also documented by the `ControlType::get` function itself, which is _exactly_ my point. If people care to know exactly what is happening under the hood the function already says it through its documentation; any semi-modern IDE will display it when typing the name of the function or hovering over it, so there is no point in repeating it.

Reply:
> And in the end correct usage of the API results in correct behavior AFAIU, nothing surprising there, so no need to point it out in a comment. 

Nothing surprising to you. I would assume most contributors will not even what the API is. The fact they may be responsible for the bugs themselves (as in this case) has little to do with whether this code should be readable by the people who will be reading it. 

> Well it certainly is not just a constructor in the C++ sense. This behavior is also documented by the ControlType::get function itself, which is exactly my point. 

Why would I assume a get function of one type affects a string pointer in a different type? The line of code in MLIR belongs to a class of a 3rd type, and also does not have a comment describing what it does. 

I do think the code design here is poor. It doesn't make sense to me that copying into the MLIR context happen when we pass the ArrayRef to the channel, it feels like an hidden side-effect that is being purposefully obscured to attach it to this other action.

Reply:
> This behavior is also documented by the ControlType::get function itself, which is exactly my point. 

This is also I think, not true. We only figured out what was happening because you pointed us to the direct line of code responsible, and gave us a rough description of what it did. None of that info you gave us is documented in the codebase.

Reply:
@shundroid @murphe67 To give some closure on this.

Memory management in MLIR is legit complicated (for me as well) for good reasons, and while most of the times (in this project) people don't have to care because MLIR "just does the memory management for you" there are cases where this is not true. Handling the extra signals is one of those cases where I had to write custom allocation code, which is why I was not so surprised when you guys thought you found a bug; this sounded very possible to me.

For future reference, this allocation logic is somewhat described in [this section of the official MLIR docs](https://mlir.llvm.org/docs/DefiningDialects/AttributesAndTypes/#parameters) (read subsection below as well). It does not go deep enough in my opinion, one piece of evidence being that I had to go look at autogen code to figure out what was really happening under the hood. Once you do that the underlying design decisions make sense but we cannot hope that people will do this, and they most likely shouldn't have to either. What I tried to convey was that properly documenting such behavior probably requires a full-fledged tutorial given that the long~ish official docs themselves are too high-level, and this just cannot fit in an inline comment. Feel free however to leave a warning for future onlookers and a link to the docs (on the function's comment itself for greater visibility).

Reply:
Yeah I see what you're saying.

An update on our side is we are have a pull request which simplifies the custom allocation code from 3 classes down to 1, and moves the allocation into c++- my issue is mostly to do with how the classes are currently structured.

I think that is a discussion for once we are satisfied with everything else here first though?

Reply:
Sounds good, thanks!